### ISSUE TITLE
Styletweaks windows: size and scrolling issue

### ISSUE THREAD

**Issue by mergen3107** [koreader/koreader#8082]
* KOReader version: 2021.07-45
* Device: Onyx Boox Kon-Tiki 2 (Android 10)

#### Issue
![Screenshot_20210816-035741](https://user-images.githubusercontent.com/6279855/129535365-b0ae3aa1-66d4-46bc-8aae-ac56b5f2b9bf.png)

1) When I long press a styles tweak file which is large, the style tweaks window height is larger than the screen height.
2) If I scroll down, I can see buttons, as on the screen. However, I am unable to scroll up. 

#### Steps to reproduce
1) Put any of these css files to %userdata/koreader/styletweaks : 
[custom css files - mergen3107.zip](https://github.com/koreader/koreader/files/6990889/custom.css.files.-.mergen3107.zip)
2) Restart KOReader, open a book
3) In `Top Menu - Doc tab - Style tweaks - User style tweaks` long press the css file.
4) Opened window is taller than the screen height.
5) Scroll down, and now you cannot scroll upon.

**Comment by poire-z** [comment]
> 1. When I long press a styles tweak file which is large, the style tweaks window height is larger than the screen height.

Not really an issue.
First, style tweaks should not be that large and fat (see https://github.com/koreader/koreader/issues/8074#issuecomment-899367773).

There shouldn't really be a need to have the text there scrollable. if needed, we could try to reduce the font size to make it all fit on the screen.
But as the window is a MovableContainer, you can still move it with long-press-and-pan to move the tall window and put parts of it into view - and see all the content, and reach the bottom buttons. It's painful, but it works.

> 2. If I scroll down, I can see buttons, as on the screen. However, I am unable to scroll up.

The behaviour you notice about "scroll down" works but "scroll up" doesn't is that when you swipe the window to move it (instead of long-press-and-pan it), it adjusts to screen boundaries so to not overflow - and it happens that we handle the bottom overflow last, so for a larger than screen window, we'll end up putting the bottom of the window to the bottom of the screen.

**Comment by mergen3107** [comment]
> when you swipe the window to move it (instead of long-press-and-pan it), it adjusts to screen boundaries so to not overflow

I understand that this is made in a way so that at least buttons can be visible with scroll down, but from the UI/UX point of view this is not very pleasant, I would say.
Maybe the text should be truncated if it force the window to increase in height, and there should be "Edit" button as well which will bring you to the Text Editor, where scrolling is handled well? 

Also, what if in Portrait view the snippet can fit into window, but the user looks at in Landscape? Then again the window will not fit.

**Comment by poire-z** [comment]
> I understand that this is made in a way so that at least buttons can be visible with scroll down

Not really, that was just a lucky side effects, not something I planned in MovableContainer :)

Anyway, long-press on style tweak is read only, it's just for looking at, shipped style tweaks are all small and can't be edited.
User style tweaks use the same display, but should also be mostly looked at and small - they can be edited thru Text Editor, but you have to find them with the file chooser here (but they won't refresh until next KOreader restart I thing).

If these style tweaks should be larger, it's probably because you are targetting a specific book with many class names, and then, you should use "Book style tweak", that opens with the TextEditor - and there, there is more handling on update while running.

**Comment by mergen3107** [comment]
> TLDR: i don't want to work on solving this issue :)

OK, I understand :) Thank you for your feedback!

By the way, in `View HTML` -> `View xx.css` (epub's own css) there is a very nice scroll-able read-only window with button at the bottom:
![Screenshot_20210824-171643](https://user-images.githubusercontent.com/6279855/130691437-4d7a3609-ee38-43ed-9e15-03a6d335ccd4.png)

Is it difficult to use the same window style for Style Tweaks?
This window is pretty much the answer to my OP.

**Comment by mergen3107** [comment]
P.S. and it is so useful, especially with `Prettify`! Thank you very much to those who implemented this so flawlessly!

**Comment by poire-z** [comment]
> Is it difficult to use the same window style for Style Tweaks?

This window is a regular TextViewer, and is fixed size. The Style tweak window fits to the (usually small) content, has a description, a long_press handled to copy the tweak to clipboard for use elsewhere, other buttons, and is pretty :)

> This window is pretty much the answer to my OP.

But your OP was for when you didn't know style tweaks were small and pretty, and not huge noisy stylesheets :) and so don't need a TextViewer.

**Comment by mergen3107** [comment]
> a long_press handled to copy the tweak to clipboard

Nice feature! Although it seems that long press handles the container movement and transparency, and to copy I just need to tap once on the text.

OK, so I managed to fit my css tweaking needs into 1 window page (so far):
![Screenshot_20210824-173233](https://user-images.githubusercontent.com/6279855/130693150-f4ece24c-6308-43d0-97e9-0072463b2f85.png)

However, I feel like I will occupy the next page too :D 

Thank you @poire-z for all your explanations regarding css!
I finally found my Zen with a good mix of default epub.css + publisher styles + Style Tweaks:
![Screenshot_20210824-173602](https://user-images.githubusercontent.com/6279855/130693262-f0394cae-f0ca-4ed0-97c9-8d8def7485da.png)

I can't just through away embedded epub style in some books, because some epubs have uniquely tailored styles, which (might) look great.

**Comment by poire-z** [comment]
> and to copy I just need to tap once on the text.

Oups, indeed, the trick was "tap", not "hold".

> I can't just through away embedded epub style in some books, because some epubs have uniquely tailored styles, which (might) look great.

I think you'd better think it the other way around: accept publisher styles - and correct/disable bits of it with style tweaks.

Your P tweaks are probably there (and if the value changes, duplicate it as a User tweak):
![image](https://user-images.githubusercontent.com/24273478/130696512-2a61b5f0-5cf7-400b-97a4-b18b15922eba.png)
If there's a single one of all of yours CSS in that file that happens to mistmach a classname (say, P.first, could be all the P in the first chapter), you're doomed: you have to disable this full tweak.

Having (nearly) each of your selectors in individual user tweaks would allow you to toggle some of them when they don't do as expected or you don't want them - and you can set them all as defaults.
Same results, but a lot cleaner and flexible.

**Comment by mergen3107** [comment]
> Having (nearly) each of your selectors in individual user tweaks would allow you to toggle some of them

Indeed, thank you! I am slowly grasping all the perks of Style Tweaks :D 

For p.first, I only applied it so that the first letter style does not get messy with p. Without this fix, the first letter gets double text-indent, and also the first line gets its indent after the first letter.

**Comment by poire-z** [comment]
Hoping I have given you the taste for small and concise style tweaks, closing :)
